[Stanford_dogs] No examples yielded

**Short description**
When performing the following snippet of code:
`tfds.load("stanford_dogs")`

the error: _AssertionError("No examples were yielded.")_ is thrown

**Environment information**
* Operating System: Windows 10
* Python version: 3.6.10 |Anaconda, Inc.| (default, Mar 23 2020, 17:58:33) [MSC v.1916 64 bit (AMD64)]
* `tensorflow-datasets`/`tfds-nightly` version: tensorflow-datasets=2.1.0
* `tensorflow`/`tensorflow-gpu`/`tf-nightly`/`tf-nightly-gpu` version:
Via _'pip freeze | findstr tensorflow'_
tensorflow==2.1.0
tensorflow-datasets==2.1.0
tensorflow-estimator==2.1.0
tensorflow-metadata==0.21.1


**Reproduction instructions**

```
import tensorflow_datasets as tfds
tfds.load("stanford_dogs")
```

**Link to logs**
_[1mDownloading and preparing dataset stanford_dogs/0.2.0 (download: 778.12 MiB, generated: Unknown size, total: 778.12 MiB) to C:\Users\Flowhill\tensorflow_datasets\stanford_dogs\0.2.0...[0m
Dl Completed...: 0 url [00:00, ? url/s]
Dl Size...: 0 MiB [00:00, ? MiB/s]
Dl Size...: 0 MiB [00:00, ? MiB/s]

Dl Completed...: 0 url [00:00, ? url/s]
Dl Completed...: 0 url [00:00, ? url/s]
Dl Size...: 0 MiB [00:00, ? MiB/s]

Extraction completed...: 0 file [00:00, ? file/s]
Extraction completed...: 0 file [00:00, ? file/s]

Dl Size...: 0 MiB [00:00, ? MiB/s]

Dl Completed...: 0 url [00:00, ? url/s]
Shuffling and writing examples to C:\Users\Flowhill\tensorflow_datasets\stanford_dogs\0.2.0.incomplete37A45O\stanford_dogs-train.tfrecord
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\Flowhill\anaconda3\envs\tf-gpu\lib\site-packages\tensorflow_datasets\core\api_utils.py", line 52, in disallow_positional_args_dec
    return fn(*args, **kwargs)
  File "C:\Users\Flowhill\anaconda3\envs\tf-gpu\lib\site-packages\tensorflow_datasets\core\registered.py", line 305, in load
    dbuilder.download_and_prepare(**download_and_prepare_kwargs)
  File "C:\Users\Flowhill\anaconda3\envs\tf-gpu\lib\site-packages\tensorflow_datasets\core\api_utils.py", line 52, in disallow_positional_args_dec
    return fn(*args, **kwargs)
  File "C:\Users\Flowhill\anaconda3\envs\tf-gpu\lib\site-packages\tensorflow_datasets\core\dataset_builder.py", line 340, in download_and_prepare
    download_config=download_config)
  File "C:\Users\Flowhill\anaconda3\envs\tf-gpu\lib\site-packages\tensorflow_datasets\core\dataset_builder.py", line 1078, in _download_and_prepare
    max_examples_per_split=download_config.max_examples_per_split,
  File "C:\Users\Flowhill\anaconda3\envs\tf-gpu\lib\site-packages\tensorflow_datasets\core\dataset_builder.py", line 931, in _download_and_prepare
    self._prepare_split(split_generator, **prepare_split_kwargs)
  File "C:\Users\Flowhill\anaconda3\envs\tf-gpu\lib\site-packages\tensorflow_datasets\core\dataset_builder.py", line 1106, in _prepare_split
    shard_lengths, total_size = writer.finalize()
  File "C:\Users\Flowhill\anaconda3\envs\tf-gpu\lib\site-packages\tensorflow_datasets\core\tfrecords_writer.py", line 211, in finalize
    self._shuffler.bucket_lengths, self._path)
  File "C:\Users\Flowhill\anaconda3\envs\tf-gpu\lib\site-packages\tensorflow_datasets\core\tfrecords_writer.py", line 88, in _get_shard_specs
    shard_boundaries = _get_shard_boundaries(num_examples, num_shards)
  File "C:\Users\Flowhill\anaconda3\envs\tf-gpu\lib\site-packages\tensorflow_datasets\core\tfrecords_writer.py", line 107, in _get_shard_boundaries
    raise AssertionError("No examples were yielded.")
AssertionError: No examples were yielded._

**Expected behavior**
Stanford dogs was already downloaded so no download bars there as expected.
Running this with tfds.load("mnist") produces no errors.

**Additional context**
A similar problem happened to a user using the PlantVillage and The300wLpTest datasets [here](https://github.com/tensorflow/datasets/issues/1670).

> 
> 
> @Flowhill It's working fine please recheck I think there is some issue with your system refer this [colab ](https://colab.research.google.com/drive/1uZa2NDJE00fpTFsky6UKVwl_cixr_vgQ) notebook.
> 
> Edit : Also check locally for `Stanford_dogs `running fine, also run fine for plant_village dataset [colab](https://colab.research.google.com/drive/1eUdhCVkjzsiNkzKTuJCOI_1AjOv3il_1) Try to reinstall TFDS` pip uninstall tensorflow_datasets`

It does indeed seem to work via colab.
Reinstalled tfds `pip uninstall tensorflow_datasets pip install tensorflow_datasets` and tried plant_village. It throws the same error.
This is done on a clean conda environment having installed tensorflow, tensorflow-gpu and pip installing tensorflow_datasets (as the anaconda cloud has an outdated version of it)

_Dl Completed...: 100%|████████████████████████████████████████████████████████████████| 1/1 [05:19<00:00, 319.00s/ url]
Shuffling and writing examples to C:\Users\Flowhill\tensorflow_datasets\plant_village\1.0.0.incomplete16HE9U\plant_village-train.tfrecord
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\Flowhill\anaconda3\envs\tf-gpu\lib\site-packages\tensorflow_datasets\core\api_utils.py", line 52, in disallow_positional_args_dec
    return fn(*args, **kwargs)
  File "C:\Users\Flowhill\anaconda3\envs\tf-gpu\lib\site-packages\tensorflow_datasets\core\registered.py", line 305, in load
    dbuilder.download_and_prepare(**download_and_prepare_kwargs)
  File "C:\Users\Flowhill\anaconda3\envs\tf-gpu\lib\site-packages\tensorflow_datasets\core\api_utils.py", line 52, in disallow_positional_args_dec
    return fn(*args, **kwargs)
  File "C:\Users\Flowhill\anaconda3\envs\tf-gpu\lib\site-packages\tensorflow_datasets\core\dataset_builder.py", line 340, in download_and_prepare
    download_config=download_config)
  File "C:\Users\Flowhill\anaconda3\envs\tf-gpu\lib\site-packages\tensorflow_datasets\core\dataset_builder.py", line 1078, in _download_and_prepare
    max_examples_per_split=download_config.max_examples_per_split,
  File "C:\Users\Flowhill\anaconda3\envs\tf-gpu\lib\site-packages\tensorflow_datasets\core\dataset_builder.py", line 931, in _download_and_prepare
    self._prepare_split(split_generator, **prepare_split_kwargs)
  File "C:\Users\Flowhill\anaconda3\envs\tf-gpu\lib\site-packages\tensorflow_datasets\core\dataset_builder.py", line 1106, in _prepare_split
    shard_lengths, total_size = writer.finalize()
  File "C:\Users\Flowhill\anaconda3\envs\tf-gpu\lib\site-packages\tensorflow_datasets\core\tfrecords_writer.py", line 211, in finalize
    self._shuffler.bucket_lengths, self._path)
  File "C:\Users\Flowhill\anaconda3\envs\tf-gpu\lib\site-packages\tensorflow_datasets\core\tfrecords_writer.py", line 88, in _get_shard_specs
    shard_boundaries = _get_shard_boundaries(num_examples, num_shards)
  File "C:\Users\Flowhill\anaconda3\envs\tf-gpu\lib\site-packages\tensorflow_datasets\core\tfrecords_writer.py", line 107, in _get_shard_boundaries
    raise AssertionError("No examples were yielded.")
AssertionError: No examples were yielded._

> 
> 
> Just do this `import glob` at top, then replace `tf.io.gfile.glob` with `glob.iglob()`

Yes it works! Thank you very much, you are a blessing!

Let me summarize:
My problem was using a fresh anaconda environment created by:
`conda create --name <name> tensorflow tensorflow-gpu`
and
`pip install tensorflow-datasets`

Error: 
_raise AssertionError("No examples were yielded.")
AssertionError: No examples were yielded._

Occurred with the following code:
`import tensorflow_datasets as tfds
tfds.load("stanford_dogs")`
or
`tfds.load("plant_village")`

The solution for plant_village is to navigate to 
~\anaconda3\envs\<name>\Lib\site-packages\tensorflow_datasets\image called plant_village.py

add `import glob` at the top and replace line 
`for fpath in tf.io.gfile.glob(glob_path):`
by
`for fpath in glob.iglob(glob_path):`

Now, my only remaining problem is that in stanford_dogs.py glob is not used. Instead, it uses multiple accounts of `tf.io.gifile.~`. Could you help me with that?

@Flowhill  Thanks you for showing results, actually its not good to use `glob.iglob `we have to use `tf.io.gfile` because to provide  GCS support. 
Are you tried to reinstall TF ? because they fix this issue, But pip package is not updated yet so you can simply use `pip install tensorflow==2.2.0rc2` and not to change `tf.io.gfile`
But its fine if you able to work with `glob.iglob()` as in future updated version we cannot have this error in TFDS

For `Stanford_dogs `  error is in `_generate_examples` so there is nothing wrong with `tf.io.gfile` but I think problem is with `_NAME_RE.match` thats why its gives  `AssertionError: No examples were yielded` maybe it not matches pattern given correctly.

> 
> 
> @Flowhill Thanks you for showing results, actually its not good to use `glob.iglob `we have to use `tf.io.gfile` because to provide GCS support.
> Are you tried to reinstall TF ? because they fix this issue, But pip package is not updated yet so you can simply use `pip install tensorflow==2.2.0rc2` and not to change `tf.io.gfile`
> But its fine if you able to work with `glob.iglob()` as in future updated version we cannot have this error in TFDS
> 
> For `Stanford_dogs ` error is in `_generate_examples` so there is nothing wrong with `tf.io.gfile` but I think problem is with `_NAME_RE.match` thats why its gives `AssertionError: No examples were yielded` maybe it not matches pattern given correctly.

Updating tensorflow using `pip install tensorflow==2.2.0rc2` solves the `tf.io.gfile` problem.
The original stanford_dogs problem is still there.
I've tried to check whether somethign goes wrong with `_NAME_RE.match` by adding a print statement printing both `_NAME_RE` and `fname`, but it seems to output these properly.
I'm going to debug a bit more to find out when it throws the error.

--- a/tensorflow_datasets/image_classification/stanford_dogs.py
+++ b/tensorflow_datasets/image_classification/stanford_dogs.py
@@ -61,7 +61,7 @@ address = "Colorado Springs, CO",
 _IMAGES_URL = "http://vision.stanford.edu/aditya86/ImageNetDogs/images.tar"
 _SPLIT_URL = "http://vision.stanford.edu/aditya86/ImageNetDogs/lists.tar"
 _ANNOTATIONS_URL = "http://vision.stanford.edu/aditya86/ImageNetDogs/annotation.tar"
-_NAME_RE = re.compile(r"([\w-]*/)*([\w]*.jpg)$")
+_NAME_RE = re.compile(r"([\w-]*[/|\\])*([\w]*.jpg)$")
 
 
 class StanfordDogs(tfds.core.GeneratorBasedBuilder):
@@ -106,7 +106,7 @@ class StanfordDogs(tfds.core.GeneratorBasedBuilder):
       with tf.io.gfile.GFile(file_name, "rb") as f:
         parsed_mat_arr = scipy.io.loadmat(f, squeeze_me=True)
       file_list = [
-          element.split("/")[-1] for element in parsed_mat_arr["file_list"]
+          os.path.split(element)[-1] for element in parsed_mat_arr["file_list"]
       ]
 
       return file_list, parsed_mat_arr
@@ -118,7 +118,7 @@ class StanfordDogs(tfds.core.GeneratorBasedBuilder):
       if "train" in fname:
         train_list, train_mat_arr = parse_mat_file(full_file_name)
         label_names = set([  # Set to remove duplicates
-            element.split("/")[-2].lower()  # Extract path/label/img.jpg
+            os.path.split(element)[-2].lower()  # Extract path/label/img.jpg
             for element in train_mat_arr["file_list"]
         ])
       elif "test" in fname:
@@ -166,7 +166,7 @@ class StanfordDogs(tfds.core.GeneratorBasedBuilder):
 
     for fname, fobj in archive:
       res = _NAME_RE.match(fname)
-      if not res or (fname.split("/")[-1] not in file_names):
+      if not res or (os.path.split(fname)[-1] not in file_names):
         continue
 
       label = res.group(1)[:-1].lower()

